Enums are not handled correctly when they are used as map keys

Example:

```java
import io.avaje.jsonb.Json;
import io.avaje.jsonb.Jsonb;

import java.util.Map;

@Json
public record EnumExample(String name, Map<Thing, String> thingMap) {

    public enum Thing {
        ONE,
        TWO
    }

    public static void main(String[] args) {
        var jsonb = Jsonb.builder().build();
        var type = jsonb.type(EnumExample.class);
        var example = new EnumExample("test", Map.of(
                EnumExample.Thing.ONE, "1",
                EnumExample.Thing.TWO, "2"
        ));
        System.out.println(type.toJson(example));
    }

}
```

That will produce an exception like:

```
Exception in thread "main" io.avaje.jsonb.JsonException: java.lang.ClassCastException: class com.pwinckles.EnumExample$Thing cannot be cast to class java.lang.String (com.pwinckles.EnumExample$Thing is in unnamed module of loader 'app'; java.lang.String is in module java.base of loader 'bootstrap')
	at io.avaje.jsonb.core.DJsonType.toJson(DJsonType.java:89)
	at io.avaje.jsonb.core.DJsonType.toJson(DJsonType.java:61)
	at com.pwinckles.EnumExample.main(EnumExample.java:23)
Caused by: java.lang.ClassCastException: class com.pwinckles.EnumExample$Thing cannot be cast to class java.lang.String (com.pwinckles.EnumExample$Thing is in unnamed module of loader 'app'; java.lang.String is in module java.base of loader 'bootstrap')
	at io.avaje.jsonb.core.MapAdapter.toJson(MapAdapter.java:51)
	at io.avaje.jsonb.core.MapAdapter.toJson(MapAdapter.java:27)
	at io.avaje.jsonb.NullSafeAdapter.toJson(NullSafeAdapter.java:33)
	at com.pwinckles.jsonb.EnumExampleJsonAdapter.toJson(EnumExampleJsonAdapter.java:52)
	at com.pwinckles.jsonb.EnumExampleJsonAdapter.toJson(EnumExampleJsonAdapter.java:11)
	at io.avaje.jsonb.NullSafeAdapter.toJson(NullSafeAdapter.java:33)
	at io.avaje.jsonb.core.DJsonType.toJson(DJsonType.java:86)
	... 2 more
```

--- a/blackbox-test/src/main/java/org/example/customer/enums/EnumExample.java
+++ b/blackbox-test/src/main/java/org/example/customer/enums/EnumExample.java
@@ -1,12 +1,12 @@
 package org.example.customer.enums;
 
+import java.util.EnumMap;
 import java.util.Map;
-
 
 import io.avaje.jsonb.Json;
 
 @Json
-public record EnumExample(String name, Map<Thing, String> thingMap) {
+public record EnumExample(String name, EnumMap<Thing, String> thingMap) {
 
   public enum Thing {
     ONE,

--- a/blackbox-test/src/main/java/org/example/customer/enums/EnumExample.java
+++ b/blackbox-test/src/main/java/org/example/customer/enums/EnumExample.java
@@ -1,0 +1,15 @@
+package org.example.customer.enums;
+
+import java.util.Map;
+
+
+import io.avaje.jsonb.Json;
+
+@Json
+public record EnumExample(String name, Map<Thing, String> thingMap) {
+
+  public enum Thing {
+    ONE,
+    TWO
+  }
+}

--- a/blackbox-test/src/test/java/org/example/customer/enums/EnumExampleTest.java
+++ b/blackbox-test/src/test/java/org/example/customer/enums/EnumExampleTest.java
@@ -2,8 +2,10 @@ package org.example.customer.enums;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+import java.util.EnumMap;
 import java.util.Map;
 
+import org.example.customer.enums.EnumExample.Thing;
 import org.junit.jupiter.api.Test;
 
 import io.avaje.jsonb.JsonType;
@@ -15,7 +17,9 @@ class EnumExampleTest {
 
   @Test
   void to_From_Json() {
-    final var bean = new EnumExample("name", Map.of(EnumExample.Thing.ONE, "two"));
+    final EnumMap<Thing, String> map = new EnumMap<>(EnumExample.Thing.class);
+    map.put(EnumExample.Thing.ONE, "two");
+    final var bean = new EnumExample("name", map);
     final var str = jsonb.toJson(bean);
     assertThat(str).isEqualTo("{\"name\":\"name\",\"thingMap\":{\"ONE\":\"two\"}}");
 

--- a/blackbox-test/src/test/java/org/example/customer/enums/EnumExampleTest.java
+++ b/blackbox-test/src/test/java/org/example/customer/enums/EnumExampleTest.java
@@ -1,0 +1,25 @@
+package org.example.customer.enums;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Map;
+
+import org.junit.jupiter.api.Test;
+
+import io.avaje.jsonb.JsonType;
+import io.avaje.jsonb.Jsonb;
+
+class EnumExampleTest {
+
+  final JsonType<EnumExample> jsonb = Jsonb.builder().build().type(EnumExample.class);
+
+  @Test
+  void to_From_Json() {
+    final var bean = new EnumExample("name", Map.of(EnumExample.Thing.ONE, "two"));
+    final var str = jsonb.toJson(bean);
+    assertThat(str).isEqualTo("{\"name\":\"name\",\"thingMap\":{\"ONE\":\"two\"}}");
+
+    final var staff = jsonb.fromJson(str);
+    assertThat(bean).isEqualTo(staff);
+  }
+}

--- a/jsonb-generator/src/main/java/io/avaje/jsonb/generator/FieldReader.java
+++ b/jsonb-generator/src/main/java/io/avaje/jsonb/generator/FieldReader.java
@@ -197,6 +197,7 @@ final class FieldReader {
       if ("java.util.List".equals(topType) || "java.util.Set".equals(topType)) {
         types.add(genericType.firstParamType());
       } else if ("java.util.Map".equals(topType)) {
+        types.add(genericType.firstParamType());
         types.add(genericType.secondParamType());
       } else {
         types.add(topType);

--- a/jsonb/src/main/java/io/avaje/jsonb/core/CoreAdapterBuilder.java
+++ b/jsonb/src/main/java/io/avaje/jsonb/core/CoreAdapterBuilder.java
@@ -44,6 +44,7 @@ final class CoreAdapterBuilder {
     this.factories.add(new MathAdapters(mathAsString));
     this.factories.add(CollectionAdapter.FACTORY);
     this.factories.add(MapAdapter.FACTORY);
+    this.factories.add(EnumMapAdapter.FACTORY);
     this.factories.add(ArrayAdapter.FACTORY);
     this.factories.add(OptionalAdapters.FACTORY);
   }

--- a/jsonb/src/main/java/io/avaje/jsonb/core/EnumMapAdapter.java
+++ b/jsonb/src/main/java/io/avaje/jsonb/core/EnumMapAdapter.java
@@ -3,12 +3,12 @@ package io.avaje.jsonb.core;
 import io.avaje.jsonb.*;
 
 import java.lang.reflect.Type;
-import java.util.LinkedHashMap;
+import java.util.EnumMap;
 import java.util.Map;
 
 /** Converts maps with Enum keys to JSON objects. */
 @SuppressWarnings("unchecked")
-final class EnumMapAdapter<K extends Enum, V> implements JsonAdapter<Map<K, V>> {
+final class EnumMapAdapter<K extends Enum<K>, V> implements JsonAdapter<Map<K, V>> {
 
   static final Factory FACTORY =
       (type, jsonb) -> {
@@ -45,11 +45,11 @@ final class EnumMapAdapter<K extends Enum, V> implements JsonAdapter<Map<K, V>> 
 
   @Override
   public Map<K, V> fromJson(JsonReader reader) {
-    final Map<K, V> result = new LinkedHashMap<>();
+    final Map<K, V> result = new EnumMap<>(enumClass);
     reader.beginObject();
     while (reader.hasNextField()) {
       final String name = reader.nextField();
-      final var enumVal = (K) Enum.valueOf(enumClass, name);
+      final var enumVal = Enum.valueOf(enumClass, name);
       final V value = valueAdapter.fromJson(reader);
       final V replaced = result.put(enumVal, value);
       if (replaced != null) {

--- a/jsonb/src/main/java/io/avaje/jsonb/core/EnumMapAdapter.java
+++ b/jsonb/src/main/java/io/avaje/jsonb/core/EnumMapAdapter.java
@@ -1,0 +1,70 @@
+package io.avaje.jsonb.core;
+
+import io.avaje.jsonb.*;
+
+import java.lang.reflect.Type;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/** Converts maps with Enum keys to JSON objects. */
+@SuppressWarnings("unchecked")
+final class EnumMapAdapter<K extends Enum, V> implements JsonAdapter<Map<K, V>> {
+
+  static final Factory FACTORY =
+      (type, jsonb) -> {
+        final Class<?> rawType = Util.rawType(type);
+        if (rawType != Map.class) {
+          return null;
+        }
+        final var types = Util.mapValueTypes(type, rawType);
+        if (!((Class<?>) types[0]).isEnum()) {
+          return null;
+        }
+        return new EnumMapAdapter<>(jsonb, types).nullSafe();
+      };
+  private final Class<K> enumClass;
+  private final JsonAdapter<V> valueAdapter;
+
+  EnumMapAdapter(Jsonb jsonb, Type[] valueType) {
+    this.enumClass = (Class<K>) valueType[0];
+    this.valueAdapter = jsonb.adapter(valueType[1]);
+  }
+
+  @Override
+  public void toJson(JsonWriter writer, Map<K, V> map) {
+    writer.beginObject();
+    for (final var entry : map.entrySet()) {
+      if (entry.getKey() == null) {
+        throw new JsonDataException("Map key is null at " + writer.path());
+      }
+      writer.name(entry.getKey().name());
+      valueAdapter.toJson(writer, entry.getValue());
+    }
+    writer.endObject();
+  }
+
+  @Override
+  public Map<K, V> fromJson(JsonReader reader) {
+    final Map<K, V> result = new LinkedHashMap<>();
+    reader.beginObject();
+    while (reader.hasNextField()) {
+      final String name = reader.nextField();
+      final var enumVal = (K) Enum.valueOf(enumClass, name);
+      final V value = valueAdapter.fromJson(reader);
+      final V replaced = result.put(enumVal, value);
+      if (replaced != null) {
+        throw new JsonDataException(
+            String.format(
+                "Map key '%s' has multiple values at path %s : %s and %s",
+                name, reader.location(), replaced, value));
+      }
+    }
+    reader.endObject();
+    return result;
+  }
+
+  @Override
+  public String toString() {
+    return "EnumMapAdapter(" + enumClass + ", " + valueAdapter + ")";
+  }
+}

--- a/jsonb/src/main/java/io/avaje/jsonb/core/EnumMapAdapter.java
+++ b/jsonb/src/main/java/io/avaje/jsonb/core/EnumMapAdapter.java
@@ -13,7 +13,7 @@ final class EnumMapAdapter<K extends Enum<K>, V> implements JsonAdapter<Map<K, V
   static final Factory FACTORY =
       (type, jsonb) -> {
         final Class<?> rawType = Util.rawType(type);
-        if (rawType != Map.class) {
+        if (rawType != EnumMap.class && rawType != Map.class) {
           return null;
         }
         final var types = Util.mapValueTypes(type, rawType);

--- a/jsonb/src/main/java/io/avaje/jsonb/core/EnumMapAdapter.java
+++ b/jsonb/src/main/java/io/avaje/jsonb/core/EnumMapAdapter.java
@@ -12,7 +12,7 @@ final class EnumMapAdapter<K extends Enum<K>, V> implements JsonAdapter<Map<K, V
 
   static final Factory FACTORY =
       (type, jsonb) -> {
-        final Class<?> rawType = Util.rawType(type);
+        final var rawType = Util.rawType(type);
         if (rawType != EnumMap.class && rawType != Map.class) {
           return null;
         }

--- a/jsonb/src/main/java/io/avaje/jsonb/core/MapAdapter.java
+++ b/jsonb/src/main/java/io/avaje/jsonb/core/MapAdapter.java
@@ -28,7 +28,7 @@ final class MapAdapter<V> implements JsonAdapter<Map<String, V>> {
 
   static final Factory FACTORY =
       (type, jsonb) -> {
-        final Class<?> rawType = Util.rawType(type);
+        final var rawType = Util.rawType(type);
         if (rawType != Map.class) {
           return null;
         }
@@ -49,7 +49,7 @@ final class MapAdapter<V> implements JsonAdapter<Map<String, V>> {
   @Override
   public void toJson(JsonWriter writer, Map<String, V> map) {
     writer.beginObject();
-    for (Map.Entry<String, V> entry : map.entrySet()) {
+    for (var entry : map.entrySet()) {
       if (entry.getKey() == null) {
         throw new JsonDataException("Map key is null at " + writer.path());
       }

--- a/jsonb/src/main/java/io/avaje/jsonb/core/MapAdapter.java
+++ b/jsonb/src/main/java/io/avaje/jsonb/core/MapAdapter.java
@@ -26,14 +26,19 @@ import java.util.Map;
  */
 final class MapAdapter<V> implements JsonAdapter<Map<String, V>> {
 
-  static final Factory FACTORY = (type, jsonb) -> {
-    Class<?> rawType = Util.rawType(type);
-    if (rawType != Map.class) {
-      return null;
-    }
-    Type valueType = Util.mapValueType(type, rawType);
-    return new MapAdapter<>(jsonb, valueType).nullSafe();
-  };
+  static final Factory FACTORY =
+      (type, jsonb) -> {
+        final Class<?> rawType = Util.rawType(type);
+        if (rawType != Map.class) {
+          return null;
+        }
+        final var valueTypes = Util.mapValueTypes(type, rawType);
+
+        if (valueTypes[0] != String.class) {
+          return null;
+        }
+        return new MapAdapter<>(jsonb, valueTypes[1]).nullSafe();
+      };
 
   private final JsonAdapter<V> valueAdapter;
 

--- a/jsonb/src/main/java/io/avaje/jsonb/core/Util.java
+++ b/jsonb/src/main/java/io/avaje/jsonb/core/Util.java
@@ -554,18 +554,18 @@ public final class Util {
    * Returns a two element array containing this map's key and value types in positions 0 and 1
    * respectively.
    */
-  static Type mapValueType(Type context, Class<?> contextRawType) {
+  static Type[] mapValueTypes(Type context, Class<?> contextRawType) {
     // Work around a problem with the declaration of java.util.Properties. That class should extend
     // Hashtable<String, String>, but it's declared to extend Hashtable<Object, Object>.
     if (context == Properties.class) {
-      return String.class;
+      return new Type[]{String.class,String.class};
     }
     Type mapType = supertype(context, contextRawType, Map.class);
     if (mapType instanceof ParameterizedType) {
       ParameterizedType mapParameterizedType = (ParameterizedType) mapType;
-      return mapParameterizedType.getActualTypeArguments()[1];
+      return mapParameterizedType.getActualTypeArguments();
     }
-    return Object.class;
+    return new Type[] {String.class, Object.class};
   }
 
   /**

